Add bookings

sashajaaa opened the pull request.

sashajaaa committed:
Commit №14.1:
feat:
-Добавлена БД;
-Добавлена реализация аренды (Booking).

refactor:
-Полный рефактор приложения.

kalitva commented on pom.xml:
```diff
@@ -54,6 +53,18 @@
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jpa</artifactId>
+		</dependency>
+		<dependency>
```
Хибернейт уже подтягивается вместе с data-jpa

kalitva commented on pom.xml:
```diff
@@ -54,6 +53,18 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jdbc</artifactId>
```
jdbc тоже

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,89 @@
+    @Query("UPDATE Booking b "
+            + "SET b.status = :status  "
+            + "WHERE b.id = :bookingId")
+    void update(BookingStatus status, Long bookingId);
```
Стоит пожалуй назвать `updateStatus`, а вообще для обновления объекта обычно используется `save`

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,89 @@
+            + "WHERE b.id = :bookingId")
+    void update(BookingStatus status, Long bookingId);
+
+    List<Booking> findAllByBookerIdOrderByStartDesc(long id);
```
Вместо определения всех этих методов можно использовать `Specification` фильтры, это необязательно но будет очень полезно знать

https://spring.io/blog/2011/04/26/advanced-spring-data-jpa-specifications-and-querydsl

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,89 @@
+            + "INNER JOIN Item i ON b.item.id = i.id "
+            + "WHERE i.ownerId = :ownerId "
+            + "ORDER BY b.start DESC")
+    List<Booking> findAllBookingsOwner(Long ownerId);
```
В большинстве случаев можно обойтись без `@Query`, например здесь - `findByOwnerId(ownerId, Sort sort)`

sashajaaa replied:
А очень плохо с Query? Нельзя так оставить?

kalitva commented on src/main/java/ru/practicum/shareit/booking/service/BookingService.java:
```diff
@@ -0,0 +1,144 @@
+        if (!ownerId.equals(userId)) {
+            throw new OperationAccessException(String.format("User with ID = %d is not the owner, no access to booking.", userId));
+        }
+        if (approve) {
```
Как насчет что-то типа

```java
BookingStatus status = approve ? APPROVED : REJECTED;
booking.setTatus(status);
bookingRepository.save(booking);
```

sashajaaa replied:
Это очень критично? Все тесты со статусами сыпаться начинают...

kalitva commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -3,19 +3,49 @@
-    private Long requestId;
+
+    @Transient
+    private ShortItemBookingDto lastBooking;
```
Никаких dto в классах ентити! Используй `Booking`

kalitva commented on src/main/java/ru/practicum/shareit/item/service/ItemService.java:
```diff
@@ -1,97 +1,156 @@
+                .collect(Collectors.toList());
+        return item.stream()
+                .map(this::updateBookings)
+                .peek((i) -> CommentMapper.toDtoList(commentRepository.findAllByItemId(i.getId())))
```
`peek` не рекомендуется в продакшен коде, почитай javadoc метода. Просто чтобы знал, можешь оставить если хочешь

kalitva commented on src/main/java/ru/practicum/shareit/user/mapper/UserMapper.java:
```diff
@@ -1,24 +1,22 @@
 import ru.practicum.shareit.user.dto.UserDto;
 import ru.practicum.shareit.user.model.User;
 
-@Component
```
Зачем удалил? Статика - зло

kalitva commented on src/main/resources/application.properties:
```diff
@@ -8,11 +8,8 @@ logging.level.org.springframework.transaction=INFO
-#---
 spring.config.activate.on-profile=ci,test
 spring.datasource.driverClassName=org.h2.Driver
 spring.datasource.url=jdbc:h2:mem:shareit
```
Желательно чтобы ты попробовал подключиться к postgres

kalitva commented on src/main/java/ru/practicum/shareit/item/service/ItemService.java:
```diff
@@ -1,97 +1,156 @@
+    @Transactional
+    public List<ItemDto> search(String text) {
         if (text == null || text.isBlank()) {
-            return Collections.emptyList();
```
Так правильнее было

sashajaaa committed:
Commit №14.2:
refactor:
-переработан файл pom.xml;
-переработан файл application.properties;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-переработан класс BookingService;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemService;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService.

sashajaaa committed:
Commit №14.2:
refactor:
-переработан файл pom.xml;
-переработан файл application.properties;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-переработан класс BookingService;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemService;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService.

sashajaaa committed:
Commit №14.2:
refactor:
-переработан файл pom.xml;
-переработан файл application.properties;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-переработан класс BookingService;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemService;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService.

sashajaaa committed:
Commit №14.2:
refactor:
-переработан файл pom.xml;
-переработан файл application.properties;
-переработан класс BookingController;
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-переработан класс BookingService;
-переработан класс CommentMapper;
-переработан класс ItemController;
-переработан класс ItemMapper;
-переработан класс Item;
-переработан класс ItemService;
-переработан класс UserController;
-переработан класс UserMapper;
-переработан класс UserService.

kalitva commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -1,16 +1,40 @@
 public class BookingMapper {
-    public static BookingDto toBookingDto(Booking booking) {
-        return BookingDto.builder()
+    public static OutputBookingDto toBookingDto(Booking booking) {
```
Сделай методы инстанс и инжекти мапперы спрингом

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,89 @@
+                                                                              LocalDateTime start,
+                                                                              BookingStatus status);
+
+    @Query("SELECT b FROM Booking b "
```
`findByOwnerId --> findByItemOwnerId` и должно заработать без `@Query`

sashajaaa replied:
не заработало

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,89 @@
+
+    List<Booking> findAllByBookerIdOrderByStartDesc(long id);
+
+    List<Booking> findAllByBookerIdAndStatusIsOrderByStartDesc(Long id, BookingStatus status);
```
Слоао `All` необязательно, обычно его не используют

kalitva commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -6,25 +6,21 @@
 public class ItemMapper {
-
-    public ItemDto toItemDto(Item item) {
+    public static ItemDto toItemDto(Item item) {
```
Без static

kalitva commented on src/main/java/ru/practicum/shareit/booking/repository/BookingRepository.java:
```diff
@@ -0,0 +1,89 @@
+    @Query("UPDATE Booking b "
+            + "SET b.status = :status  "
+            + "WHERE b.id = :bookingId")
+    void save(BookingStatus status, Long bookingId);
```
Не вижу особого смысла переопределять родительский `save`

sashajaaa committed:
Commit №14.3:
refactor:
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-переработан класс BookingService;
-переработан класс CommentMapper;
-переработан класс ItemMapper;
-переработан класс ItemService;
-переработан класс UserMapper;
-переработан класс UserService.

sashajaaa committed:
Commit №14.3:
refactor:
-переработан класс BookingMapper;
-переработан интерфейс BookingRepository;
-переработан класс BookingService;
-переработан класс CommentMapper;
-переработан класс ItemMapper;
-переработан класс ItemService;
-переработан класс UserMapper;
-переработан класс UserService.
